lib/carbon/cache.py

```diff
@@ -112,23 +112,28 @@ def _generate_queue():
 
     self.queue = _generate_queue()
 
-  def choose_item(self):
+  def choose_item(self, mdpu=False):
     return next(self.queue)
 
 
 class TimeSortedStrategy(DrainStrategy):
-  """ This strategy prefers metrics wich are lagging behind
+  """ This strategy prefers metrics which are lagging behind
   guarantees every point gets written exactly once during
   a loop of the cache """
   def __init__(self, cache):
     super(TimeSortedStrategy, self).__init__(cache)
+    self.mdpu_flag = False
 
     def _generate_queue():
       while True:
         t = time.time()
         metric_lw = sorted(self.cache.watermarks, key=lambda x: x[1], reverse=True)
         if settings.MIN_TIMESTAMP_LAG:
           metric_lw = [x for x in metric_lw if t - x[1] > settings.MIN_TIMESTAMP_LAG]
+        if self.mdpu_flag and settings.MIN_DATAPOINTS_PER_UPDATE > 0:
+          metric_counts = [x for x in sorted(self.cache.counts, key=lambda x: x[1])
+                           if x[1] >= settings.MIN_DATAPOINTS_PER_UPDATE]
+          metric_lw.append(metric_counts)
```

Comment:
Here don't we want the intersection of the 2 lists? otherwise anything that has passed the min timestamp lag will be written regardless of the number of points.  For that to work I'd change both list comprehensions to return `x[0] for x in`, then you can use `metric_lw = list(set(metric_lw) & set(metric_counts))`, and remove the use of `itemgetter(0)` on line 144.

Reply:
That's a good quiestion. @Exocomp asked for "greedy" mechanics here - but now I'm wondering does it have any sense...

Reply:
Should work now, but now I'm afraid that's could be quite slow. Anyway, I don't think that we should merge this w/o thorough test under load, which is quite non-trivial.

Reply:
actually we could speed it up with:
```
metric_lw = [x[0] for x in sorted(self.cache.counts, key=lambda x: x[1])
                  if x[1] >= settings.MIN_DATAPOINTS_PER_UPDATE and x[0] in metric_lw]
```

Reply:
Yes, but that way breaks an order of metric_lw list. :(

Reply:
ah ok, so we want to keep the metrics_lw order, then we have to have 2 statements:
```
metric_count = [x[0] for x in self.cache.counts if x[1] >= settings.MIN_DATAPOINTS_PER_UPDATE]
metric_lw = [x for x in metric_lw if x in metric_count]
```
no need to sort the cache counts since we don't care about the order just whether they're over the threshold

Reply:
Ah, indeed - that looks much better! Thanks!